src/main/java/jenkins/scm/api/trait/SCMHeadAuthority.java

```diff
@@ -0,0 +1,212 @@
+ * context of a specific {@link SCMSourceRequest}.
+ * <p>
+ * Note: there can be multiple authorities for the same types of head / revision active in the context of any one
+ * request. The ultimate trust state is determined by a logical OR operation, in other words if any one authority
+ * says that the head / revision is trusted then the head / revision is trusted.
+ *
+ * @param <S> the type of {@link SCMSourceRequest}.
+ * @param <H> the specialization of {@link SCMHeadMixin} that this authority provides trust information for.
+ * @param <R> the specialization of {@link SCMRevision} that this authority provides trust information for.
+ * @since 3.4.0
+ */
+public abstract class SCMHeadAuthority<S extends SCMSourceRequest, H extends SCMHeadMixin, R extends SCMRevision>
+        extends AbstractDescribableImpl<SCMHeadAuthority<?, ?, ?>> {
+
+    /**
+     * Checks if this authority is relevant to the supplied {@link SCMHead}.
+     *
+     * @param head the supplied {@link SCMHead}.
+     * @return {@code true} if and only if this authority can provide trust information for the supplied head.
+     */
+    public final boolean isApplicableTo(@NonNull SCMHead head) {
+        return getDescriptor().isApplicableToHead(head);
+    }
+
+    /**
+     * Checks if this authority is relevant to the supplied {@link SCMRevision}.
+     *
+     * @param revision the supplied {@link SCMRevision}.
+     * @return {@code true} if and only if this authority can provide trust information for the supplied revision.
+     */
+    public final boolean isApplicableTo(@NonNull SCMRevision revision) {
+        return getDescriptor().isApplicableToRevision(revision);
+    }
+
+    /**
+     * Checks if this authority is relevant to the supplied {@link SCMSourceRequest}.
+     *
+     * @param request the supplied {@link SCMSourceRequest}.
+     * @return {@code true} if and only if this authority can provide trust information for the supplied requesy.
+     */
+    public final boolean isApplicableTo(@NonNull SCMSourceRequest request) {
+        return getDescriptor().isApplicableToRequest(request);
+    }
+
+    /**
+     * Checks if the supplied {@link SCMHead} is trusted in the context of the specified {@link SCMSourceRequest}.
+     *
+     * @param request the {@link SCMSourceRequest}.
+     * @param head    the {@link SCMHead}.
+     * @return {@code true} if the supplied head is trusted.
+     * @throws IOException          if there was an I/O error trying to establish the trust status.
+     * @throws InterruptedException if interrupted while trying to establing the trust status.
+     */
+    @SuppressWarnings("unchecked")
+    public final boolean isTrusted(@NonNull SCMSourceRequest request, @NonNull SCMHead head)
+            throws IOException, InterruptedException {
+        return isApplicableTo(request)
+                && isApplicableTo(head)
+                && checkTrusted((S) request, (H) head);
```

Comment:
Ahh but they won't!

`isApplicableTo(request)` is `getDescriptor().isApplicableTo(request)` which ultimately ensures that the request is at least an instance of the parameterized type in the implementation's constructor.

same goes for `isApplicableTo(head)` so the casts will always work.

However that does raise the point that implementers should ensure they call the base method when overriding, hence 0b6a184e37bfc552b4636a13d94639e1756bc53f